src/core/layout/qgslayoutitemchart.cpp

```diff
@@ -0,0 +1,390 @@
+    return;
+
+  if ( mLayout->renderContext().isPreviewRender() )
+  {
+    if ( mNeedsGathering || mIsGathering )
+    {
+      if ( mNeedsGathering )
+      {
+        mNeedsGathering = false;
+        refreshData();
+      }
+
+      QgsScopedQPainterState painterState( painter );
+      painter->setClipRect( thisPaintRect );
+
+      painter->setBrush( QBrush( QColor( 125, 125, 125, 125 ) ) );
+      painter->drawRect( thisPaintRect );
+      painter->setBrush( Qt::NoBrush );
+      QFont messageFont;
+      messageFont.setPointSize( 12 );
+      painter->setFont( messageFont );
+      painter->setPen( QColor( 255, 255, 255, 255 ) );
+      painter->drawText( thisPaintRect, Qt::AlignCenter | Qt::AlignHCenter, tr( "Rendering chart" ) );
+      return;
+    }
+  }
+  else
+  {
+    if ( mNeedsGathering )
+    {
+      mNeedsGathering = false;
+      prepareGatherer();
+      QgsApplication::instance()->taskManager()->addTask( mGatherer.data() );
+      mGatherer->waitForFinished( 60000 );
+    }
+  }
+
+  const double scaleFactor = QgsLayoutUtils::scaleFactorFromItemStyle( itemStyle, painter );
+  const QSizeF size = mLayout->convertToLayoutUnits( sizeWithUnits() ) * scaleFactor;
+  if ( size.width() == 0 || size.height() == 0 )
+    return;
+
+  mPlot->setSize( size );
+
+  QgsScopedQPainterState painterState( painter );
+  painter->scale( 1 / scaleFactor, 1 / scaleFactor );
+
+  QgsRenderContext renderContext = QgsLayoutUtils::createRenderContextForLayout( mLayout, painter );
+  renderContext.setScaleFactor( scaleFactor );
+  renderContext.setExpressionContext( createExpressionContext() );
+
+  QgsPlotRenderContext plotRenderContext;
+  mPlot->render( renderContext, plotRenderContext, mPlotData );
+}
+
+void QgsLayoutItemChart::refresh()
+{
+  QgsLayoutItem::refresh();
+  mNeedsGathering = true;
+}
+
+void QgsLayoutItemChart::refreshData()
+{
+  mGathererTimer.start();
+}
+
+void QgsLayoutItemChart::gatherData()
+{
+  prepareGatherer();
+  QgsApplication::instance()->taskManager()->addTask( mGatherer.data() );
+
+  mIsGathering = true;
+  update();
+}
+
+void QgsLayoutItemChart::prepareGatherer()
+{
+  if ( mGatherer )
+  {
+    disconnect( mGatherer.data(), &QgsTask::taskCompleted, this, &QgsLayoutItemChart::processData );
+    mGatherer->cancel();
+    mGatherer.clear();
+  }
+
+  if ( !mVectorLayer || !mPlot || mSeriesList.isEmpty() )
+  {
+    mPlotData.clearSeries();
+    mIsGathering = false;
+    update();
+  }
+
+  QList<QgsVectorLayerXyPlotDataGatherer::XySeriesDetails> xYSeriesList;
+  for ( const SeriesDetails &series : mSeriesList )
+  {
+    xYSeriesList << QgsVectorLayerXyPlotDataGatherer::XySeriesDetails( series.xExpression(), series.yExpression(), series.filterExpression() );
+  }
+
+  if ( Qgs2DXyPlot *xyPlot = dynamic_cast<Qgs2DXyPlot *>( mPlot.get() ) )
+  {
+    QgsFeatureRequest request;
+    if ( mSortFeatures && !mSortExpression.isEmpty() )
+    {
+      request.addOrderBy( mSortExpression, mSortAscending );
+    }
```

Comment:
you could combine the filter expressions here and set a filter on the request for optimal performance on large layers